quarkusio/quarkus
# Build time analytics

This PR enables an **opt-in** feature to allow gathering of build time analytics to get info/insights about how Quarkus is being used during development/builds (only build time NOT runtime).

# TO TEST THIS 

build the PR first using `mvn -Dquickly`, then create a project using latest snapshot and with a non quarkus grouped, for example: `quarkus create app -P io.quarkus:quarkus-bom:999-SNAPSHOT something.random:myapp`.

Please try and use your own groupid/name so we get more diverse reports.

Then run `quarkus dev` in the project and say `y` to report data.
Running in dev mode and mvn quarkus:dev, and gradle quarkusDev should report then submit data on each run.

Check `target/build-analytics-event.json` for content of what was sent.

# Implementation
It is implemented using [Segment](https://segment.com/) which is used by other Red Hat projects to gather anonymous usage data.

Intent is that on first quarkus dev mode user will be asked if they wish to share anonymously build time info to Quarkus team.

If answer is no, marker file is created in `~/.redhat` and no recording/analytics is made before that markerfile is removed.

If answer is yes, markerfile is made + `~/.redhat/anonymousId` is created which is just to be able to group data per installation without any identifiable info in them. From that point on when ever quarkus build plugin runs it will send a json packet with info. This is sent async and non-blocking. Meaning build won't block because of this.

* One-way SHA-265 hashes representing the application:
   * groupId+artifactId
   * artifactId+version
* Name/Version info for Java, GraalVM, Operating System, Build system
* Whether or not Kubernetes is used on the build
* Known Extensions enabled 
    i.e. group-ids "io.quarkus", "io.quarkiverse", “org.apache.camel.quarkus” or listed in Quarkus  platform. 
* The IP is set to 0.0.0.0 on purpose to disable any possible inference, if not present.

An example of data collected is shown below - all non-identifiable info.

```
{
  "context": {
    "app": {
      "name": "ZzOlt3q0VgNK0oyUd6NDAhPqJjTSxNOAnqHb5dHfS7M=",
      "version": "FUDdh16tTKtuJkcoY2Wp57IKKwxZ/Hx42tTvi3Izpmc="
    },
    "build": {
      "gradle_version": "N/A",
      "maven_version": "3.8.7"
    },
    "ci": {
      "name": "unknown"
    },
    "graalvm": {
      "java_version": "N/A",
      "vendor": "N/A",
      "version": "N/A"
    },
    "ip": "0.0.0.0",
    "java": {
      "vendor": "Eclipse Adoptium",
      "version": "17.0.4.1"
    },
    "kubernetes": {
      "detected": "false"
    },
    "location": {
      "country": "PT"
    },
    "os": {
      "name": "Mac OS X",
      "os_arch": "aarch64",
      "version": "13.3.1"
    },
    "quarkus": {
      "version": "999-SNAPSHOT"
    },
    "timezone": "Europe/Lisbon"
  },
  "event": "PROD_BUILD",
  "integrations": {},
  "messageId": "api-2QCEFlcxqJeuKjX4Uf0m1aWoKED",
  "originalTimestamp": "2023-05-23T13:37:39.913591Z",
  "properties": {
    "app_extensions": [
      {
        "artifact_id": "quarkus-resteasy-jackson",
        "group_id": "io.quarkus",
        "version": "999-SNAPSHOT"
      },
      {
        "artifact_id": "quarkus-smallrye-openapi",
        "group_id": "io.quarkus",
        "version": "999-SNAPSHOT"
      },
      {
        "artifact_id": "quarkus-hibernate-orm",
        "group_id": "io.quarkus",
        "version": "999-SNAPSHOT"
      },
      {
        "artifact_id": "quarkus-opentelemetry",
        "group_id": "io.quarkus",
        "version": "999-SNAPSHOT"
      },
      {
        "artifact_id": "quarkus-jdbc-postgresql",
        "group_id": "io.quarkus",
        "version": "999-SNAPSHOT"
      },
      {
        "artifact_id": "quarkus-arc",
        "group_id": "io.quarkus",
        "version": "999-SNAPSHOT"
      }
    ]
  },
  "receivedAt": "2023-05-23T13:37:40.856Z",
  "timestamp": "2023-05-23T13:37:39.913Z",
  "type": "track",
  "userId": "58991333-5f25-4401-ae5b-e3eb09ceb4de"
}
```
